benchmarks/replay/replay.cpp

```diff
@@ -74,42 +79,128 @@ struct allocation {
+  {
+  }
+
+  /**
+   * @brief Copy construct a replay_benchmark
+   *
+   * Does not copy the mutex or the map
+   */
+  replay_benchmark(replay_benchmark const& other)
+    : factory_{other.factory_},
+      mr_{other.mr_},
+      events_{other.events_},
+      allocation_map{events_.size()},
+      event_index{0}
+  {
+  }
+
+  /// Add an allocation to the map (thread safe)
+  void set_allocation(std::unordered_map<uintptr_t, allocation>& allocation_map,
+                      uintptr_t ptr,
+                      allocation alloc)
+  {
+    std::lock_guard<std::mutex> lock(map_mutex);
+    allocation_map.insert({ptr, alloc});
+  }
+
+  /// Remove an allocation from the map (thread safe)
+  allocation remove_allocation(std::unordered_map<uintptr_t, allocation>& allocation_map,
+                               uintptr_t ptr)
   {
+    std::lock_guard<std::mutex> lock(map_mutex);
+    auto iter = allocation_map.find(ptr);
+    if (iter != allocation_map.end()) {
+      allocation a = iter->second;
+      allocation_map.erase(iter);
+      return a;
+    }
+    return allocation{};
+  }
+
+  /// Create the memory resource shared by all threads before the benchmark runs
+  void SetUp(const ::benchmark::State& state) override
+  {
+    if (state.thread_index == 0) {
+      rmm::logger().log(spdlog::level::info, "------ Start of Benchmark -----");
+      mr_ = factory_();
+    }
   }
 
-  void operator()(benchmark::State& state)
+  /// Destroy the memory resource and count any unallocated memory
+  void TearDown(const ::benchmark::State& state) override
   {
-    // Maps a pointer from the event log to an active allocation
-    std::unordered_map<uintptr_t, allocation> allocation_map(events_.size());
+    if (state.thread_index == 0) {
+      rmm::logger().log(spdlog::level::info, "------ End of Benchmark -----");
+      // clean up any leaked allocations
+      std::size_t total_leaked{0};
+      std::size_t num_leaked{0};
+      for (auto const& ptr_alloc : allocation_map) {
+        auto alloc = ptr_alloc.second;
+        num_leaked++;
+        total_leaked += alloc.size;
+        mr_->deallocate(alloc.p, alloc.size, 0);
+      }
+      if (num_leaked > 0)
+        std::cout << "LOG shows leak of " << num_leaked << " allocations of " << total_leaked
+                  << " total bytes\n";
+      allocation_map.clear();
+      mr_.reset();
+    }
+  }
 
+  /// Run the replay benchmark
+  void BenchmarkCase(::benchmark::State& state) override
+  {
     auto const& my_events = events_.at(state.thread_index);
 
+    using namespace std::chrono_literals;
+
     for (auto _ : state) {
-      std::for_each(my_events.begin(), my_events.end(), [&allocation_map, &state, this](auto e) {
+      std::for_each(my_events.begin(), my_events.end(), [&state, this](auto e) {
+        // ensure correct ordering between threads
+        while (event_index < e.index) {
+          std::this_thread::sleep_for(2us);
+        }
```

Comment:
My suspicion is that it would be better to use a `std::condition_variable` here.

Reply:
Something like this: https://godbolt.org/z/rxqGWE

Reply:
Thanks!  Done.